proof/refine/ARM/KHeap_R.thy

```diff
@@ -3795,5 +3795,23 @@ lemma cross_relF:
   "(s, s') \<in> state_relation \<Longrightarrow> cross_rel A B \<Longrightarrow> A s \<Longrightarrow> B s'"
   by (clarsimp simp: cross_rel_def)
 
+lemma valid_simple'[simp]:
+  "valid_tcb_state' Running = \<top>"
+  "valid_tcb_state' Inactive = \<top>"
+  "valid_tcb_state' Restart = \<top>"
+  "valid_tcb_state' IdleThreadState = \<top>"
+  by (rule ext, simp add: valid_tcb_state'_def)+
+
+lemma tcb_at'_ex1_ko_at':
+  "tcb_at' t s \<Longrightarrow> \<exists>!tcb. ko_at' (tcb::tcb) t s"
+  by (clarsimp simp: obj_at'_def)
+
+(* FIXME RT: move to Lib? *)
```

Comment:
I think the pure logic lemma could go to lib because I think it's probably a fundamental logical identity. I managed to prove the following stronger lemma actually: `Ex1 Q = (∀P. (∃x. Q x ∧ P x) = (∀x. Q x ⟶ P x))`. So saying "there exists a unique x such that Q x" is equivalent to saying that Q interacts with all other propositions in the above manner. And that interaction is "P implies Q if and only if P and Q is true somewhere."

We don't have ot move it to lib, but it would good to move it somewhere fairly high up.

Reply:
But `KHeap_R` is fairly high up, so happy to leave it here.

Reply:
I've been thinking about this more. The reason that this rewrite is so useful is that the left hand side has standard intro rules which `clarsimp` will apply in the goal, but not the hypotheses. And the right hand side has standard elimination rules which `clarsimp` will apply in the hypotheses, but not the goal. So adding this to the simp set (or its symmetric version) will effectively apply standard elimination/introduction rules without the need for manual intervention.